src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseAndTransferService.java

```diff
@@ -848,6 +852,71 @@ public InmateDetail courtTransferIn(final String offenderNo, final RequestForCou
         return offenderTransformer.transform(offenderBooking);
     }
 
+    public InmateDetail temporaryAbsenceArrival(final String offenderNo, final RequestForTemporaryAbsenceArrival requestForTemporaryAbsenceArrival) {
+
+        final OffenderBooking offenderBooking = this.getOffenderBooking(offenderNo);
+        if (!offenderBooking.getInOutStatus().equals("OUT")) {
+            throw new BadRequestException("Prisoner is not currently out");
+        }
+        final var latestExternalMovement = offenderBooking.getLastMovement().map(lm -> {
+            if (!lm.getMovementType().getCode().equals(TAP.getCode())) {
+                throw new BadRequestException("Latest movement not a temporary absence");
+            }
+
+            if (!lm.isActive()) {
+                throw new BadRequestException("Temporary absence transfer not active");
+            }
+            return lm;
+        }).orElseThrow(EntityNotFoundException.withMessage("No movements found"));
+
+        if (!requestForTemporaryAbsenceArrival.getAgencyId().equals(latestExternalMovement.getFromAgency().getId())) {
+            throw new BadRequestException("Prisoner is not returning to the same prison");
+        }
+
+        offenderBooking.setInOutStatus(IN.name());
+        offenderBooking.setLivingUnitMv(null);
+
+        deactivatePreviousMovements(offenderBooking);
+
+        final MovementReason movementReason = getMovementReason(requestForTemporaryAbsenceArrival.getMovementReasonCode(), latestExternalMovement.getMovementReason());
+
+        final LocalDateTime movementTime = getAndCheckMovementTime(requestForTemporaryAbsenceArrival.getDateTime(), offenderBooking.getBookingId());
+
+        final ExternalMovement.ExternalMovementBuilder builder = ExternalMovement.builder()
+            .movementDate(movementTime.toLocalDate())
+            .movementTime(movementTime)
+            .movementType(movementTypeRepository.findById(TAP).orElseThrow(EntityNotFoundException.withMessage(format("No %s movement type found", MovementType.TAP))))
+            .movementReason(movementReason)
+            .movementDirection(MovementDirection.IN)
+            .fromAgency(latestExternalMovement.getToAgency())
```

Comment:
I think we should add .fromCity(latestExternalMovement.getToCity())

Reply:
done

Reply:
done